app/master/build_scheduler.py

```diff
@@ -0,0 +1,122 @@
+from queue import Empty
+from threading import Lock
+
+from app.master.slave import SlaveMarkedForShutdownError
+from app.util import analytics
+from app.util.log import get_logger
+
+# pylint: disable=protected-access
+# Disable protected-access for whole file; will be fixing this (by
+# moving Build's subjob queues into this class) in a separate change.
+
+
+class BuildScheduler(object):
+    """
+    This class handles the logic of taking a Build and distributing its subjobs to
+    slaves. There is a one-to-one relationship between instances of this class and
+    instances of Build.
+
+    The data flow between this class and Build is unidirectional; the build instance
+    itself shouldn't know anything about its scheduler or slaves or where subjobs
+    are executing. All that goes here.
+
+    Ususally this class is instantiated and managed by a BuildSchedulerPool.
+    """
```

Comment:
Almost everything in this class is pulled unchanged from the original methods in `Build`.